source/tutorials/learning-journey/packaging-existing-software.md

```diff
@@ -0,0 +1,531 @@
+
+:::{note}
+`callPackage` automatically passes attributes from `nixpkgs` to the given function, if they match attributes required by that function's argument attrset.
+
+In this case, `callPackage` will supply `lib`, and `stdenv` to the function defined in `hello.nix`.
+:::
+
+Now run the `nix-build` command with the new argument:
+
+```console
+$ nix-build -A hello
+error: derivation name missing
+```
+
+This new failure occurs with the *derivation*, further down in the file than the initial error on line 2 about the `lib` argument not having a value; the previous error was successfully resolved by changing the expression passed to `nix-build`.
+
+### Naming a Derivation
+Every derivation needs a `name` attribute, which must either be set directly or constructed by `mkDerivation` from `pname` and `version` attributes, if they exist.
+
+Update the file again to add a `name`:
+
+```nix
+# hello.nix
+{ lib
+, stdenv
+, fetchzip
+}:
+
+stdenv.mkDerivation {
+  name = "hello";
+
+  src = fetchzip {
+    url = "https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz";
+    sha256 = lib.fakeSha256;
+  };
+}
+```
+and then re-run the command:
+
+```console
+$ nix-build -A hello
+error:
+...
+       … while evaluating attribute 'src' of derivation 'hello'
+
+         at /home/nix-user/hello.nix:9:3:
+
+            8|
+            9|   src = fetchzip {
+             |   ^
+           10|     url = "https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz";
+
+       error: hash mismatch in file downloaded from 'https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz':
+         specified: sha256:0000000000000000000000000000000000000000000000000000
+         got:       sha256:0xw6cr5jgi1ir13q6apvrivwmmpr5j8vbymp0x6ll0kcv6366hnn
+```
+
+### Finding The File Hash
+As expected, the incorrect file hash caused an error, and Nix helpfully provided the correct one, which you can now substitute into `hello.nix` to replace `lib.fakeSha256`:
+
+```nix
+# hello.nix
+{ lib
+, stdenv
+, fetchzip
+}:
+
+stdenv.mkDerivation {
+  name = "hello";
+
+  src = fetchzip {
+    url = "https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz";
+    sha256 = "0xw6cr5jgi1ir13q6apvrivwmmpr5j8vbymp0x6ll0kcv6366hnn";
+  };
+}
+```
+
+Now run the previous command again:
+
+```console
+$ nix-build -A hello
+this derivation will be built:
+  /nix/store/rbq37s3r76rr77c7d8x8px7z04kw2mk7-hello.drv
+building '/nix/store/rbq37s3r76rr77c7d8x8px7z04kw2mk7-hello.drv'...
+...
+configuring
+...
+configure: creating ./config.status
+config.status: creating Makefile
+...
+building
+... <many more lines omitted>
+```
+Great news: the derivation built successfully!
+
+The console output shows that `configure` was called, which produced a `Makefile` that was then used to build the project; it wasn't necessary to write any build instructions in this case, because the `stdenv` build system is based on `autoconf`, which automatically detected the structure of the project directory.
```

Comment:
```suggestion
The console output shows that `configure` was called, which produced a `Makefile` that was then used to build the project.
It wasn't necessary to write any build instructions in this case because the `stdenv` build system is based on `autoconf`, which automatically detected the structure of the project directory.
```